### ISSUE TITLE
Finer tuning of exceptions for warnings

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#692]
In GitLab by @bsipocz on Mar 5, 2019, 15:01

We would like to enlarge the variety of warnings we are checking for our codebase in CI. Having the ability for a finer tuning of the exceptions would be super useful.

E.g. the warnings for `F401` are reporting all our ``__init__.py`` files. It would be super useful to be able to ignore imports from the same package rather than skipping them line by line.

E.g. in our case, anything from astropy is OK in those files, but everyone else should result the warning when unused.

**Comment by asottile** [comment]
In GitLab by @bsipocz on Mar 5, 2019, 15:07

Or, maybe in this special case, ignoring relative imports in specific files would solve my issue.

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 5, 2019, 15:14

`per-file-ignores` was added in `flake8>=3.7` which should satisfy this usecase.

For example, ignoring F401 in all `__init__.py` files:

```ini
[flake8]
per-file-ignores =
    */__init__.py: F401
```

**Comment by asottile** [comment]
In GitLab by @bsipocz on Mar 5, 2019, 15:16

Yes, but it ignore all import warnings there. Our issue is that we  would like to  still receive warnings for the ones that are not imports of the given package (or even submodule).

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 5, 2019, 15:21

Reminder that `flake8` is just a framework which wraps some tools, it adds `# noqa` as a feature but itself doesn't differentiate lines as import vs code vs etc.  The various codes produced by flake8 come from the tools it wraps (in this case, pyflakes)

your best bet is to either:

1. expose the names you want to expose in `__all__ = ('name', 'name', 'name')` (this will exempt them from unused import checking from pyflakes)
2. `noqa` the lines you want to avoid lint errors from

**Comment by asottile** [comment]
In GitLab by @bsipocz on Mar 5, 2019, 16:10

Good point that I should have raised this with pyflakes, thanks for the comments.

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 5, 2019, 16:12

pyflakes has no options so you probably won't get far there either, I'd suggest using `__all__` personally

**Comment by asottile** [comment]
In GitLab by @bsipocz on Mar 5, 2019, 16:19

`__all__` is not practical at all, simply there are just too many stuff in that namespace level and `__all__`s are used one layer down, e.g. https://github.com/astropy/astropy/blob/master/astropy/coordinates/__init__.py

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 5, 2019, 16:21

then it looks like `# noqa` on each of those lines is your only option.

**Comment by asottile** [comment]
In GitLab by @bsipocz on Mar 5, 2019, 16:22

yes, thus the feature request :)

**Comment by asottile** [comment]
In GitLab by @bsipocz on Mar 5, 2019, 16:24

(But I don't have a good idea what kind of configuration setup would be generic enough to be useful for more warning types.)

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 5, 2019, 16:30

I don't think flake8 will add something like this, it's way too specific and unlikely to be helpful for any beyond astropy.

You ~could probably write an astropy specific formatter plugin which filters out errors that occur in `__init__.py` files and involve `.` relative imports, but I don't see how something that specific could ever make it into core

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Mar 6, 2019, 05:19

The problem with the formatter avenue is that we would still exit non-zero even if the formatter decided to not print any of those warnings... and if one uses another formatter for machine-readable output then it would show those errors when the custom one here doesn't.

That doesn't mean we need a solution here, but it does point out what Astropy will encounter

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 6, 2019, 09:03

Ah true, maybe a wrapper script might work then

Something like:

```python
import subprocess
import sys


def filename_line(s: bytes) -> bytes:
    return b':'.join(s.split(b':')[:2])


def main() -> int:
    grep_cmd = ('git', 'grep', '-n', r'from \.', '--', '*/__init__.py')
    grep = subprocess.check_output(grep_cmd)
    exclude_lines = {filename_line(s) for s in grep.splitlines()}
    flake8_cmd = ('flake8', *sys.argv[1:])
    flake8 = subprocess.run(flake8_cmd, stdout=subprocess.PIPE)
    ret = 0
    for line in flake8.stdout.splitlines(True):
        if filename_line(line) not in exclude_lines:
            sys.stdout.buffer.write(line)
            ret = 1
    return ret


if __name__ == '__main__':
    exit(main())
```

```console
$ flake8 astropy/coordinates/builtin_frames/__init__.py --select=F
astropy/coordinates/builtin_frames/__init__.py:44:1: F401 '.icrs_fk5_transforms' imported but unused
astropy/coordinates/builtin_frames/__init__.py:45:1: F401 '.fk4_fk5_transforms' imported but unused
astropy/coordinates/builtin_frames/__init__.py:46:1: F401 '.galactic_transforms' imported but unused
astropy/coordinates/builtin_frames/__init__.py:47:1: F401 '.supergalactic_transforms' imported but unused
astropy/coordinates/builtin_frames/__init__.py:48:1: F401 '.icrs_cirs_transforms' imported but unused
astropy/coordinates/builtin_frames/__init__.py:49:1: F401 '.cirs_observed_transforms' imported but unused
astropy/coordinates/builtin_frames/__init__.py:50:1: F401 '.intermediate_rotation_transforms' imported but unused
astropy/coordinates/builtin_frames/__init__.py:51:1: F401 '.ecliptic_transforms' imported but unused
$ python3 flake8-astropy.py astropy/coordinates/builtin_frames/__init__.py --select=F
$
```

**Comment by asottile** [comment]
In GitLab by @asottile on Aug 18, 2019, 15:43

Since I don't think there's anything actionable in core here I'm closing -- there's some sample code above that should do ~roughly what you want (but may require some tweaking)

**Comment by asottile** [comment]
In GitLab by @asottile on Aug 18, 2019, 15:43

closed
